pkg/discovery/grpcresolver/grpc_resolver_test.go

```diff
@@ -0,0 +1,178 @@
+	testInstance := &test{}
+	for i := 0; i < count; i++ {
+		lis, err := net.Listen("tcp", "localhost:0")
+		assert.NoError(t, err, "failed to listen on tcp")
+		s := grpc.NewServer()
+		grpctest.RegisterTestServiceServer(s, &testServer{})
+		testInstance.servers = append(testInstance.servers, s)
+		testInstance.addresses = append(testInstance.addresses, lis.Addr().String())
+
+		go func(s *grpc.Server, l net.Listener) {
+			s.Serve(l)
+		}(s, lis)
+	}
+
+	return testInstance
+}
+
+func makeSureConnectionsUp(t *testing.T, count int, testc grpctest.TestServiceClient) {
+	var p peer.Peer
+	addrs := make(map[string]struct{})
+	// Make sure connections to all servers are up.
+	for si := 0; si < count; si++ {
+		connected := false
+		for i := 0; i < 1000; i++ {
+			_, err := testc.EmptyCall(context.Background(), &grpctest.Empty{}, grpc.Peer(&p))
+			if err != nil {
+				continue
+			}
+			if _, ok := addrs[p.Addr.String()]; !ok {
+				addrs[p.Addr.String()] = struct{}{}
+				connected = true
+				break
+			}
+			time.Sleep(time.Millisecond * 10)
+		}
+		assert.True(t, connected, "Connection was still not up")
+	}
+}
+
+func assertRoundRobinCall(t *testing.T, connections int, testc grpctest.TestServiceClient) {
+	addrs := make(map[string]struct{})
+	var p peer.Peer
+	for i := 0; i < connections; i++ {
+		_, err := testc.EmptyCall(context.Background(), &grpctest.Empty{}, grpc.Peer(&p))
+		assert.NoError(t, err)
+		addrs[p.Addr.String()] = struct{}{}
+	}
+	assert.Len(t, addrs, connections, "must call each of the servers once")
+}
+
+func TestErrorDiscoverer(t *testing.T) {
+	notifier := &discovery.Dispatcher{}
+	errMessage := errors.New("error discoverer returns error")
+	discoverer := erroredDiscoverer{
+		err: errMessage,
+	}
+	r := New(notifier, discoverer, zap.NewNop(), 2)
+	_, err := r.Build(resolver.Target{}, nil, resolver.BuildOption{})
+	assert.Equal(t, errMessage, err)
+}
+
+func TestGRPCResolverRoundRobin(t *testing.T) {
+	backendCount := 5
+
+	testInstances := startTestServers(t, backendCount)
+	defer testInstances.cleanup()
+
+	notifier := &discovery.Dispatcher{}
+	discoverer := discovery.FixedDiscoverer{}
+
+	tests := []struct {
+		minPeers    int
+		connections int
+	}{
+		{3, 5}, {5, 5}, {7, 5},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("minPeers=%d", test.minPeers), func(t *testing.T) {
+			res := New(notifier, discoverer, zap.NewNop(), test.minPeers)
+			defer resolver.UnregisterForTesting(res.Scheme())
+
+			cc, err := grpc.Dial(res.Scheme()+":///round_robin", grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name))
+			assert.NoError(t, err, "could not dial using resolver's scheme")
+			defer cc.Close()
+
+			testc := grpctest.NewTestServiceClient(cc)
+
+			notifier.Notify(testInstances.addresses)
+
+			connections := test.minPeers
+			if connections > test.connections {
+				connections = test.connections
+			}
+
+			// This extra step is necessary to make sure servers are ready for roundrobin test.
```

Comment:
// This step is necessary to ensure that connections to all min-peers are ready, 
// otherwise round-robin may loop only through already connected peers.